### PR TITLE
Allow non-equijoin filters in join condition

### LINKED ISSUE #659: Join condition should allow non-equijoin expressions
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Currently datafusion explicitly disallows join conditions not looking like Col1 = Col2.

We should be more flexible by moving unsupported parts of the condition to a filter instead.

**Describe the solution you'd like**
Any expression that can't be in the join condition, should move to a filter.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Dandandan**
# Which issue does this PR close?

Closes #659

 # Rationale for this change

In queries, some expressions can be added that can not be handled by a equi-join. Those should be supported by moving them to a filter instead. 
This is only done for inner join for now, supporting this for other (left/right) joins will require a bit more work.

# What changes are included in this PR?

Changes the planner to add the unsupported expressions to a filter instead of returning an error.
This almost makes TPC-H query 13 pass, except from some small unsupported / ignored syntax.

# Are there any user-facing changes?

No, should just be a bit more flexible.

### CODE REVIEW

File: datafusion/src/sql/planner.rs [hunk 1]
```diff
@@ -1556,32 +1575,32 @@ fn remove_join_expressions(
 /// foo = bar
 /// foo = bar AND bar = baz AND ...
 ///
-fn extract_join_keys(expr: &Expr, accum: &mut Vec<(Column, Column)>) -> Result<()> {
+fn extract_join_keys(
```
Comment (houqp) [inline datafusion/src/sql/planner.rs:1578]: i think the function doc string above needs to be updated :)

**Review by houqp — APPROVED** [review]
:+1:

**Comment by Dandandan** [comment]
@alamb do you mind to take a look?

File: datafusion/src/sql/planner.rs [hunk 2]
```diff
@@ -368,15 +368,34 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
                     keys.into_iter().unzip();
                 // return the logical plan representing the join
-                LogicalPlanBuilder::from(left)
-                    .join(right, join_type, left_keys, right_keys)?
+                let join = LogicalPlanBuilder::from(left)
+                    .join(right, join_type, left_keys, right_keys)?;
+
+                if filter.is_empty() {
+                    join.build()
+                } else if join_type == JoinType::Inner {
```
Comment (alamb) [inline datafusion/src/sql/planner.rs:385]: 👍  this is the right behavior.

File: datafusion/src/sql/planner.rs [hunk 3]
```diff
@@ -368,15 +368,34 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+                    join.filter(
```
Comment (alamb) [inline datafusion/src/sql/planner.rs:386]: this is a fancy way of creating an `AND` chain 👍

File: datafusion/tests/sql.rs [hunk 4]
```diff
@@ -1694,6 +1694,17 @@ async fn equijoin() -> Result<()> {
     Ok(())
 }
 
+#[tokio::test]
+async fn equijoin_unsuppored_condition() -> Result<()> {
```
Comment (alamb) [inline datafusion/tests/sql.rs:1698]: this test appears to be mis-named now. Perhaps something like

```suggestion
async fn equijoin_and_other_condition() -> Result<()> {
```

**Review by alamb — APPROVED** [review]
Nice work @Dandandan  -- the only other thing I would suggest is a negative case

Something like this with a LEFT join:

```sql
SELECT t1_id, t1_name, t2_name FROM t1 LEFT JOIN t2 ON t1_id = t2_id AND t2_name >= 'y' ORDER BY t1_id
```
And expect that that will error with unsupported

**Pull request merged**